content/posts/2017-05-03-normandy-and-auth0.mkd

```diff
@@ -0,0 +1,188 @@
+title: Switching Normandy to use Auth0
+date: 2016-05-03
+category: posts
+tags: [mozilla, normandy]
+---
+
+Recently [Normandy][] switched from authenticating ourselves users
+with boring username and passwords to using Auth0 to authenticate
+users more securely.
+
+Normandy is a web service that holds a lot of influence over
+Firefox. Because of this, we have had [a list of security
+features][sec-checklist] we've been working through. One of the big
+items on this list was to not store passwords, and do authentication
+of users ourselves.
+
+We chose to use Auth0 for this, primarily because it is the
+new hotness as far as authenticating Mozillians. It can use many
+sources of authentication, including Mozilla's LDAP servers, the
+canonical source of employee user data. This is exactly what we want
+to use for Normandy.
+
+
+# Overview
+
+Normandy is a Django app, so we initially explored doing the
+integration with Auth0 directly in the app. The idea would be to use
+an existing OpenIDC library to authenticate users with Auth0, and
+correlate that to the existing users of the system via email address.
+
+Unfortunately, we weren't able to get any of the libraries to
+work for us. The major problems we ran into were incompatibilities with
+something in our stack (Python 3.6 or Auth0 specifically) or the
+implementation being too complex.
+
+Instead we chose an easier process. Normandy is fronted by Nginx,
+which does some work with caching and logging. Our operations team
+has a Auth0 integration that works with our Nginx frontend. It passes
+authentication details to our app via the HTTP header
+`Remote-User`. This solution was much easier to implement: essentially
+we flipped a flag in puppet, and we started getting authentication
+headers.
+
+
+# Changes to Django
+
+Of course, sending the headers isn't enough. We also have to configure
+the app to read those headers and act accordingly. We did this with
+Django's [`RemoteUserBackend`][]. This works by adding a middleware
+that annotates all requests with information about the authentication
+header, and an authentication backend that reads that information to
+sign a user in or out. If a user is authenticated via the
+`Remote-User` header, but does not exist in the database, the backend
+automatically creates the user and signs them in.
+
+The default settings worked well for us. The only modification we
+needed was to tie it into our logging and settings systems.  A
+simplified version of the changes is to add `RemoteUserMiddleware` to
+the `MIDDLEWARE` setting, and adding `RemoteUserBackend` to
+`AUTHENTICATION_BACKENDS`. You can see the full
+changes [in this pull request][normandy#569].
+
+
+# Changes to Nginx
+
+To send this header to Normandy, we modified the Nginx configurations to
+perform authentication via OpenIDC with Auth0. That was implemented
+with [lua-resty-openidc][].
+
+When a request comes in to Nginx, it checks if the request has cookies
```

Comment:
Does this happen for any request at all or is it requests that match the location specified in the .conf file that's covered by OIDC authentication?